### ISSUE TITLE
fullrelease sneakily skips PyPI uploads

### ISSUE THREAD

**Issue by mgedmin** [zestsoftware/zest.releaser#199]
Recently `fullrelease` stopped uploading stuff to PyPI for me.

```
mg@platonas: ~/src/zopefoundation/zope.testrunner [git:master $>] $ fullrelease 
INFO: Starting prerelease.
Do you want to run check-manifest? (Y/n)? 
lists of files in version control and sdist match
Changelog entries for version 4.6.0:

4.6.0 (unreleased)
==================

- Make the ``subunit`` support purely optional: applications which have
  been getting the dependencies via ``zope.testrunner`` should either add
  ``zope.testrunner[subunit]`` to their ``install_requires`` or else
  depend directly on ``python-subunit``.

- New option ``--ignore-new-thread=<regexp>`` to suppress "New thread(s)"
  warnings.
Enter version [4.6.0]: 
INFO: Set setup.py's version to u'4.6.0'
INFO: Changed version from 4.6.0.dev0 to 4.6.0
INFO: History file CHANGES.rst updated.
INFO: The 'git diff':

diff --git a/CHANGES.rst b/CHANGES.rst
index bc1a7a6..b5d959c 100644
--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,7 +1,7 @@
 zope.testrunner Changelog
 *************************
 
-4.6.0 (unreleased)
+4.6.0 (2016-12-28)
 ==================
 
 - Make the ``subunit`` support purely optional: applications which have
diff --git a/setup.py b/setup.py
index acdf826..ddf53c5 100644
--- a/setup.py
+++ b/setup.py
@@ -16,7 +16,7 @@
 # When developing and releasing this package, please follow the documented
 # Zope Toolkit policies as described by this documentation.
 ##############################################################################
-version='4.6.0.dev0'
+version = '4.6.0'
 
 import os
 import sys


OK to commit this (Y/n)? 
INFO: [master b8ba230] Preparing release 4.6.0
 2 files changed, 2 insertions(+), 2 deletions(-)

INFO: Starting release.
Tag needed to proceed, you can use the following command:
git tag 4.6.0 -m "Tagging 4.6.0"
Run this command (Y/n)? 

Check out the tag (for tweaks or pypi/distutils server upload) (Y/n)? 
INFO: Doing a checkout...
Note: checking out '4.6.0'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

git checkout -b <new-branch-name>

HEAD is now at b8ba230... Preparing release 4.6.0
INFO: Tag checkout placed in /tmp/zope.testrunner-4.6.0-nrd5fM/gitclone
INFO: Changing to sub directory in tag checkout: /tmp/zope.testrunner-4.6.0-nrd5fM/gitclone
INFO: Making a source distribution of a fresh tag checkout (in /tmp/zope.testrunner-4.6.0-nrd5fM/gitclone).
Showing first few lines...
running sdist
running egg_info
creating src/zope.testrunner.egg-info
writing requirements to src/zope.testrunner.egg-info/requires.txt
writing src/zope.testrunner.egg-info/PKG-INFO
...
Showing last few lines...
adding 'zope.testrunner-4.6.0/src/zope.testrunner.egg-info/entry_points.txt'
adding 'zope.testrunner-4.6.0/src/zope.testrunner.egg-info/pbr.json'
adding 'zope.testrunner-4.6.0/src/zope.testrunner.egg-info/top_level.txt'
removing 'zope.testrunner-4.6.0' (and everything under it)

Showing first few lines...
running sdist
running egg_info
creating src/zope.testrunner.egg-info
writing requirements to src/zope.testrunner.egg-info/requires.txt
writing src/zope.testrunner.egg-info/PKG-INFO
...
Showing last few lines...
adding 'zope.testrunner-4.6.0/src/zope.testrunner.egg-info/entry_points.txt'
adding 'zope.testrunner-4.6.0/src/zope.testrunner.egg-info/pbr.json'
adding 'zope.testrunner-4.6.0/src/zope.testrunner.egg-info/top_level.txt'
removing 'zope.testrunner-4.6.0' (and everything under it)

WARNING: You must have a properly configured .pypirc file in your home dir to upload to a package index.
INFO: Starting postrelease.
Current version is 4.6.0
Enter new development version ('.dev0' will be appended) [4.6.1]: 
```

The WARNING is easy to miss in the copious output.  It should be an ERROR, and zest.releaser should ask me whether I want to continue with the postrelease steps despite that.

Additionally, I've no idea what's wrong with my `.pypirc`, which looks like this:

```ini
[pypi]
username = mgedmin
password = secret
```

`twine upload` accepts it just fine.

**Comment by mauritsvanrees** [comment]
If `twine` accepts it, *we* should accept it. Which version of `twine` are you using? I am wondering if the latest version will not accept it either, but that is just a guess.

The [Python Packaging User Guide](https://packaging.python.org/distributing/?highlight=pypirc#uploading-your-project-to-pypi) says you should use a `distutils` section with an `index-servers` option:

```
[distutils]
index-servers=pypi

[pypi]
repository = https://upload.pypi.org/legacy/
username = <username>
password = <password>
```

BTW, I see that in my own `pypirc` I have a colon (`:`) instead of an equals sign, apparently both are fine.

I think we recently removed support for old-style `pypirc` files, though I am not sure in which version that was. Probably 6.6.5 or 6.7. Well, it seems not. Looking more closely, I see that the basis actually has not changed. During release we parse the pypirc and call [`is_pypi_configured`](https://github.com/zestsoftware/zest.releaser/blob/6.7.1/zest/releaser/pypi.py#L153-L178). That code has not changed the last few years. You can see there that we either expect a `server-login` section with a `username` option (old style), or a `distutils` section with an `index-servers` option (new style).

Are you sure you have not changed your `~/.pypirc`?

As for warning versus error with a question: tricky is that `zest.releaser` can also be used to just create a tag and not a distribution. And it may not even be a Python package that you release: it works fine for a buildout with a changelog file and a `version.txt`. So uploading may not be wanted by all. Or at least not necessarily to PyPI: there are zest.releaser extensions that use `scp` to copy a distribution to a simple file server.
But then: once you are in this part of the code, it is very likely that you *do* want to upload. So we may want to change this, yes.

**Comment by mauritsvanrees** [comment]
Latest `twine` still accepts your simple config indeed. Should be easy enough to do something similar to [what they do](https://github.com/pypa/twine/blob/1.8.1/twine/utils.py#L65-L70).

**Comment by mgedmin** [comment]
> Recently fullrelease stopped uploading stuff to PyPI for me
...
> Are you sure you have not changed your ~/.pypirc?

Could be!  I got a new, larger SDD a few months ago hand did a fresh Ubuntu reinstall, and I think I then had to generate a new `.pypirc` because I forgot to copy it over (and it's not included in my regular backup, because it contains secrets which are properly kept in my encrypted passwords file).

I don't mind zest.releaser insisting on strict requirements for a `.pypirc`, but I'd like it to do the insisting in a way that tells me what I'm doing wrong and what I should do to resolve it (just printing that link to the packaging guide would suffice).  And, also, in a way that makes me notice there's a problem.

(Funny story -- this was the second time `fullrelease` skipped the upload that made me file this bug.  I didn't realize that's what had happened the 1st time, as I already had closed that terminal tab when I went to PyPI and failed to see the release I'd made.  I thought it was a server-side glitch at first, dug out the checkout in /tmp, tried a twine upload, expecting it to fail with a duplicate file/version error, and only when that succeeded I started suspecting zest.releaser and watched it more closely next time I was making a release.)
